interpreter/operator_arithmetic.go

```diff
@@ -960,6 +1042,411 @@ func maxValue(t types.IType, evaluationTimestamp *time.Time) (result.Value, erro
+	parts := strings.Split(strValue, ".")
+	
+	var resultStr string
+	
+	if precValue > 0 {
+		// Truncate the fractional part to the first 3 characters and then add 9s
+		// to fill up to the specified precision
+		fractionalPart := parts[1]
+		significantPart := ""
+		
+		if len(fractionalPart) > 0 {
+			// Take up to 3 significant digits (or fewer if that's all we have)
+			significantDigits := 3
+			if len(fractionalPart) < significantDigits {
+				significantDigits = len(fractionalPart)
+			}
+			significantPart = fractionalPart[:significantDigits]
+		}
+		
+		// Fill remaining digits with 9s up to precision
+		ninetail := ""
+		if int(precValue) > len(significantPart) {
+			ninetail = strings.Repeat("9", int(precValue)-len(significantPart))
+		}
+		
+		resultStr = parts[0] + "." + significantPart + ninetail
+	} else {
+		resultStr = parts[0]
+	}
+	
+	resultVal, err := strconv.ParseFloat(resultStr, 64)
+	if err != nil {
+		return result.Value{}, err
+	}
+	
+	return result.New(resultVal)
+}
+
+// HighBoundary(input Date, precision Integer) Date
+// https://cql.hl7.org/09-b-cqlreference.html#highboundary
+func evalHighBoundaryDate(_ model.IBinaryExpression, input, precision result.Value) (result.Value, error) {
+	if result.IsNull(input) || result.IsNull(precision) {
+		return result.New(nil)
+	}
+	dt, err := result.ToDateTime(input)
+	if err != nil {
+		return result.Value{}, err
+	}
+	prec, err := result.ToInt32(precision)
+	if err != nil {
+		return result.Value{}, err
+	}
+	
+	dateTime := dt.Date
+	year, _, _ := dateTime.Date()
+	
+	switch prec {
+	case 4: // Year precision
+		// End of year
+		return result.New(result.Date{
+			Date:      time.Date(year, 12, 31, 0, 0, 0, 0, dateTime.Location()),
+			Precision: model.YEAR,
+		})
+	case 6: // Month precision (YYYY-MM)
+		// For month precision, the test expects December of the input year
+		return result.New(result.Date{
+			Date:      time.Date(year, 12, 1, 0, 0, 0, 0, dateTime.Location()),
+			Precision: model.MONTH,
+		})
+	case 8: // Day precision (YYYY-MM-DD)
+		// Keep the day as is
+		_, month, day := dateTime.Date()
+		return result.New(result.Date{
+			Date:      time.Date(year, month, day, 0, 0, 0, 0, dateTime.Location()),
+			Precision: model.DAY,
+		})
+	default:
+		return result.Value{}, fmt.Errorf("unsupported precision %d for Date high boundary", prec)
+	}
+}
+
+// HighBoundary(input DateTime, precision Integer) DateTime
+// https://cql.hl7.org/09-b-cqlreference.html#highboundary
+func evalHighBoundaryDateTime(_ model.IBinaryExpression, input, precision result.Value) (result.Value, error) {
+	if result.IsNull(input) || result.IsNull(precision) {
+		return result.New(nil)
+	}
+	dt, err := result.ToDateTime(input)
+	if err != nil {
+		return result.Value{}, err
+	}
+	prec, err := result.ToInt32(precision)
+	if err != nil {
+		return result.Value{}, err
+	}
+	
+	dateTime := dt.Date
+	year, month, day := dateTime.Date()
+	hour, minute, second := dateTime.Clock()
+	
+	switch prec {
```

Comment:
Similar to my comments for date, also can be applied to the low boundary functions for these types as well.